compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -1147,14 +1249,37 @@ multiPackageBuildEffect relativize mPkgConfig multiPackageConfig projectOpts opt
     (IDE.initialise rule (DummyLspEnv diagnosticsLogger) (toIdeLogger loggerH) noopDebouncer (toCompileOpts opts) vfs)
     IDE.shutdown
     $ \ideState -> runActionSync ideState
-      $ case mRootPkgData of
-          Nothing -> void $ uses_ BuildMulti $ toNormalizedFilePath' <$> mpPackagePaths multiPackageConfig
-          Just (rootPkgPath, _) -> void $ use_ BuildMulti rootPkgPath
-
-data AssistantRunner = AssistantRunner { runAssistant :: FilePath -> [String] -> IO ()}
-
--- Stores a mapping from dar path to project path
-data BuildableDataDeps = BuildableDataDeps { getDataDepSource :: FilePath -> Maybe FilePath }
+      $ case (mRootPkgData, buildMode) of
+          (Just (rootPkgPath, _), _) -> void $ use_ BuildMulti rootPkgPath
+          (_, MultiPackage buildAll buildComposite) -> do
+            let compositeDars = case buildComposite of
+                  SelectCompositeDars dars -> dars
+                  AllCompositeDars -> mpCompositeDars multiPackageConfig
+                packagePaths = if getMultiPackageBuildAll buildAll
+                  then toNormalizedFilePath' <$> mpPackagePaths multiPackageConfig
+                  else nubOrd $ concatMap (fmap toNormalizedFilePath' . cdPackages) compositeDars
+            void $ uses_ BuildMulti packagePaths
+            liftIO $ traverse_ (buildAndWriteCompositeDar loggerH buildableDataDeps) compositeDars
```

Comment:
This call and its behaviour feels like something that could live in Shake, but this PR is big already so best to put into another PR